File: packages/core/src/intervals.ts
import type { PitchClass } from './types';
import { mod12, toPcSet } from './pcset';

export type IntervalVector = [number, number, number, number, number, number];

export function intervalVector(pcs: PitchClass[]): IntervalVector {
  const sorted = toPcSet(pcs);
  const vec: IntervalVector = [0, 0, 0, 0, 0, 0];
  for (let i = 0; i < sorted.length; i++) {
    for (let j = i + 1; j < sorted.length; j++) {
      const diff = mod12(sorted[j]! - sorted[i]!);
      const ic = diff <= 6 ? diff : 12 - diff;
      if (ic >= 1 && ic <= 6) {
        vec[ic - 1]!++;
      }
    }
  }
  return vec;
}

export function myhillProperty(pcs: PitchClass[]): boolean {
  const sorted = toPcSet(pcs);
  const n = sorted.length;
  if (n < 2) return false;
  for (let genericInterval = 1; genericInterval < n; genericInterval++) {
    const specificSizes = new Set<number>();
    for (let i = 0; i < n; i++) {
      const j = (i + genericInterval) % n;
      const diff = mod12(sorted[j]! - sorted[i]!);
      specificSizes.add(diff);
    }
    if (specificSizes.size !== 2) return false;
  }
  return true;
}

export function zRelated(a: PitchClass[], b: PitchClass[]): boolean {
  const va = intervalVector(a);
  const vb = intervalVector(b);
  const sameVector = va.every((v, i) => v === vb[i]);
  if (!sameVector) return false;
  const sa = toPcSet(a);
  const sb = toPcSet(b);
  if (sa.length !== sb.length) return false;
  for (let n = 0; n < 12; n++) {
    const transposed = toPcSet(sa.map(pc => pc + n));
    if (transposed.length === sb.length && transposed.every((v, i) => v === sb[i])) return false;
    const inverted = toPcSet(sa.map(pc => n - pc));
    if (inverted.length === sb.length && inverted.every((v, i) => v === sb[i])) return false;
  }
  return true;
}

File: packages/core-swift/Sources/ChrometriaCore/Intervals.swift
import Foundation

/// Interval-class vector: six counts, for interval classes 1 through 6.
///
/// Mirrors `packages/core/src/intervals.ts` exactly.
public func intervalVector(_ pcs: [PitchClass]) -> [Int] {
    let sorted = toPcSet(pcs)
    var vec = [0, 0, 0, 0, 0, 0]
    for i in 0..<sorted.count {
        for j in (i + 1)..<sorted.count {
            let diff = mod12(sorted[j] - sorted[i])
            let ic = diff <= 6 ? diff : 12 - diff
            if ic >= 1 && ic <= 6 { vec[ic - 1] += 1 }
        }
    }
    return vec
}

/// Myhill's property: every generic interval maps to exactly two distinct
/// specific (semitone) intervals.
///
/// Mirrors `packages/core/src/intervals.ts` exactly, including that a set of
/// fewer than 2 pitch classes is defined as `false` rather than vacuously true.
public func myhillProperty(_ pcs: [PitchClass]) -> Bool {
    let sorted = toPcSet(pcs)
    let n = sorted.count
    if n < 2 { return false }
    for genericInterval in 1..<n {
        var specificSizes = Set<Int>()
        for i in 0..<n {
            let j = (i + genericInterval) % n
            let diff = mod12(sorted[j] - sorted[i])
            specificSizes.insert(diff)
        }
        if specificSizes.count != 2 { return false }
    }
    return true
}

/// Two sets are Z-related if they share an interval vector but are not
/// related by transposition or inversion.
///
/// Mirrors `packages/core/src/intervals.ts` exactly.
public func zRelated(_ a: [PitchClass], _ b: [PitchClass]) -> Bool {
    let va = intervalVector(a)
    let vb = intervalVector(b)
    if va != vb { return false }
    let sa = toPcSet(a)
    let sb = toPcSet(b)
    if sa.count != sb.count { return false }
    for n in 0..<12 {
        let transposed = toPcSet(sa.map { $0 + n })
        if transposed.count == sb.count && transposed == sb { return false }
        let inverted = toPcSet(sa.map { n - $0 })
        if inverted.count == sb.count && inverted == sb { return false }
    }
    return true
}
